Improving sympy.printing.pretty.stringpict API

mmatera opened the pull request:
#### References to other Issues or PRs

I am working on giving Mathics (https://github.com/Mathics3/mathics-core/pull/1155) the ability to produce prettyprint
-like output from Mathics `Expression` objects. To avoid duplicating efforts, I tried to hook on the `sympy.printing.pretty` package, but I found some problems. 
After trying to use the `PrettyPrinter`, I realized that since `mathics.core.Expression` objects behave differently than Sympy objects, the best would be to generate text blocks directly from the `sympy.printing.pretty.stringpict.prettyForm` class. A draft of it can be seen at https://github.com/Mathics3/mathics-core/pull/1162

In this journey, I found some issues in the design of `sympy.printing.pretty.stringpict` that makes me difficult to use it directly on my project. 
This PR includes the changes I need to make it work. It also probably helps with #20894.

#### Brief description of what is fixed or changed

In this PR I propose some changes in the `stringPict` API, which I hope will make it easier to use it both inside Sympy  or in in other projects, like Mathics.



  * the methods  `subindex`, `superindex`  and `subsuperindices` were added to the `stringPict` interface as high level methods to deal with indices and subindices.
  * the methods `above`, `below`, `left`, and `right` were modified to return `prettyForm` or `stringPict`
   objects depending on the class of the object over which they are called.
   * `above`, `below`, `left`,  and `right`  now support a keyword argument `align` that allow to control the relative alignment of the elements.
   * The method `parenthesis` was added, with the same functionality of `parens`, but following the pattern in `above`, `below`, `left` and `right`. `parens` is left for backward compatibility.
   * the `root` method was fixed, and used to simplify the code in `PrettyPrinter._print_nth_root()`.
   * doctests for the modified methods were added.
   * tests and the use of the modified methods were adjusted along the code.
  
The idea is that `stringPict.[above|below,left|right|subindex|superindex|subsuperindices](...)` work as high-level methods to build `stringPict` and `prettyForm` complex objects, leaving `stringPict.[stack|next]()` methods as low-level methods. High-level methods then return objects of the same type as the caller object (`stringPict|prettyForm`), while low-level methods return tuples of `(str, int)`. 
The use of the high-level methods allows us to avoid the idiom `prettyForm(*stringPict.method(...))`, which is spread all over Sympy. This behavior seems to be suggested by the implementation of `stringPict.root`, a method that didn't work in the master branch but does it here. 

Another change included is to support horizontal and vertical alignment in text blocks, which was marked as a TODO task, and which also helps to build complex stringPicts without explicitly computing the baselines on the building routines. 

The implementation of `stringPict.root`, which is working now, suggests that this organization was considered before.

#### Other comments

I tried to modify the rest of the code as little as possible to play with my project without breaking tests. I have more suggestions for improving the pretty module, but could be go in another PR.


#### Release notes

<!-- BEGIN RELEASE NOTES -->

* printing
  * the methods  `subindex`, `superindex`  and `subsuperindices` were added to the `stringPict` interface as high level methods to deal with indices and subindices.
  * the methods `above`, `below`, `left`, and `right` were modified to return `prettyForm` or `stringPict`
   objects depending on the class of the object over which they are called.
   * `above`, `below`, `left`,  and `right`  now support a keyword argument `align` that allow to control the relative alignment of the elements.
   * The method `parenthesis` was added, with the same functionality of `parens`, but following the pattern in `above`, `below`, `left` and `right`. `parens` is left for backward compatibility.
   * the `root` method was fixed, and used to simplify the code in `PrettyPrinter._print_nth_root()`.
   * doctests for the modified methods were added.
   * tests and the use of the modified methods were adjusted along the code.
   * TODO comments were included to point out how to continue.
   * Changelog-like comments in module docstrings were removed.
<!-- END RELEASE NOTES -->

mmatera committed:
* Add align attribute in `stack`, `top` and `bottom`, `left` and `right` methods.
* `paren`, `top`, `bottom`, `left` and `right` produce `stringPict` objects
* adding `superscript` and `subscript` methods
* fixing `root`

mmatera committed:
use stringPict.root to draw n-roots

adjust comments on root

oscarbenjamin commented:
The changes here seem reasonable to me.

What I'm not sure about is the idea that downstream code would use stringpict/prettyform directly.

Consider for example that if anyone is already using these in downstream code then the changes here would break that downstream code...

I have considered making deeper changes in this code to improve performance because e.g. printing large matrices is very slow and I think that this is due to algorithmic problems that don't scale linearly. I think a major problem is the way that many operations compute strings for subexpressions eagerly rather than building up a symbolic representation of the displayed expression and deferring rendering until the end.

mmatera commented:
@oscarbenjamin, I am also considering just copying the modules to my project, but it would require maintaining them. Also, mathics-core already has Sympy as a dependency so it would be an unnecessary redundancy. Finally, for us, it would be good that the formatting of Sympy and Mathics PrettyForm look the same. The only thing we need is that the interface (or at least a part of it) to be more or less stable.

oscarbenjamin commented:
> The only thing we need is that the interface (or at least a part of it) to be more or less stable.

Which part of the interface exactly?

Do the changes here affect any of the documented interfaces:
https://docs.sympy.org/latest/modules/printing.html#sympy.printing.pretty.stringpict.stringPict

mmatera commented:
> Do the changes here affect any of the documented interfaces:

Yep, the ones I mentioned in the list:

    * the methods subscript and superscript were added to the stringPict interface
    * the methods above, below, left, right and parens were modified to return prettyForm or stringPict
    objects depending on the class of the object over which are called.
    * above, below, left, and right now support a keyword argument align that allows to control the relative alignment of the elements.
    * the root method was fixed, and used to simplify the code in PrettyPrinter._print_nth_root().

The change in the return value type of `above`, `below`, `left`,  `parens` and `right` is not fully necessary, but helps me to write more clear code. 

The other changes are necessary to ensure that equivalent expressions formatted from Mathics and from Sympy looks the same.

mmatera commented on sympy/physics/quantum/cg.py:
```diff
@@ -226,16 +226,16 @@ def _pretty(self, printer, *args):
-        s = stringPict('C' + ' '*pad)
-        s = prettyForm(*s.below(bot))
-        s = prettyForm(*s.above(top))
+            top = top.right(' '*(pad - top.width()))
```
@oscarbenjamin the most disruptive changes I need to pass all the tests are like this one: instead of calling the constructor of `prettyForm` over the result of `right` (and the other methods) I just pick the output of the method.

mmatera commented:
> We should try to check if it looks like any other downstream code is using these things that would be broken by the changes here.

Regarding this, the changes are not too disruptive, and -I guess- probably welcome.

> Consider for example that if anyone is already using these in downstream code then the changes here would break that downstream code...

I do not know if this low-level code is used somewhere else outside Sympy. But I guess is like any other change in the interface. In Mathics we have to handle a few in the recent past.

mmatera commented:
> 
> I have considered making deeper changes in this code to improve performance because e.g. printing large matrices is very slow and I think that this is due to algorithmic problems that don't scale linearly. I think a major problem is the way that many operations compute strings for subexpressions eagerly rather than building up a symbolic representation of the displayed expression and deferring rendering until the end.

Regarding of this, I have also some ideas, in particular to build "grid/matrix-like" objects. I planned to implement them on the Mathics-core side, but I could share the code here.

oscarbenjamin commented:
What is not clear to me is what exactly it is that are the incompatible changes here. You have changed all the code that uses `prettyForm` but if you had not changed all that code then what is it that would not work any more?

I think if I understand then currently prettyForm.right returns a tuple but here it is changed to return a prettyForm instead. There are alternatives that we could use that would be less disruptive to any downstream code that uses this:

1. Maybe prettyForm could behave like the tuple that was previously returned (i.e. with `__getitem__`, `__iter__`)
2. Alternative methods like `rightp` could be added instead that return prettyForm objects.

oscarbenjamin commented:
I think it is potentially fine to make incompatible changes here because I doubt that much code uses stringPict or prettyForm directly outside of the sympy codebase itself. I just want to understand what it is that is being changed and also if mathics or anything else is going to depend on what I would usually consider to be an internal interface then it needs to be recorded somewhere what exactly is being depended on.

oscarbenjamin commented:
Sometimes I look at this to find code that users something:
https://grep.app/search?q=stringPict

mmatera commented:
> Sometimes I look at this to find code that users something: https://grep.app/search?q=stringPict

prettyForm is not used neither.

oscarbenjamin commented:
I think the changes here are good to merge but will wait to see if anyone else wants to review.

> I have also some ideas, in particular to build "grid/matrix-like" objects. I planned to implement them on the Mathics-core side, but I could share the code here.

Yes, that would be good. I can't remember why now but I think that the current implementation of this here is inefficient.

asmeurer commented on sympy/printing/pretty/stringpict.py:
```diff
@@ -7,9 +7,13 @@
+    - adding the v_align() method
+    - Methods `parens`, `left`, `right`, `above` and `below` now returns
+      objects of the caller object class.
+    - The method `root` now works.
```
We should probably just get rid of this docstring. We don't really do these sorts of inline changelogs in SymPy.

mmatera replied:
OK. Gone.

asmeurer commented:
I have my own ideas on how to make the pretty printer a little better at https://github.com/sympy/sympy/issues/20894. Basically, the pretty printer should keep track of the way the expressions were built up, so that you basically have a symbolic representation of the blocks of printed expressions. This would make reasoning about certain more complex printing situations easier because you could look inside what has already been printed. So I would consider whether the APIs proposed here could be adapted to that (if not now, then at some point in the future).

Another suggestion I have here is to use operator overloading for left, right, above, and below, like maybe `a + b` and `a/b`. Although I'm uncertain if this might cause confusion with actual mathematical expressions. At the same time, something like [this](https://github.com/sympy/sympy/pull/27257/files#diff-3fd23f9a6c4768c78481c24d66a4810ea7ec1ca3b81bfe78159405eb6cb3c3a5R174) is very confusing, particularly because the `left` method is "backwards".

oscarbenjamin commented:
Yes, exactly. Ideally the way that display printing should work is that an expression is first converted to a new type say DisplayExpression which represents how the expression should appear i.e. it symbolically represents things like brackets, superscripts etc. In the conversion you would have `Add -> DisplayAdd`, `Mul -> DisplayFraction/DisplayMul`, `Pow -> DisplaySurd/DisplayPow` etc. A function like `sin(x)` would be transformed to `DisplayCall('sin', x)` and so on. Then the various display printers like mathml, latex, pretty etc should operate on DisplayExpression and render the actual text. Then all Mathics would need to do is implement the transformation from a Mathics expression to DisplayExpression but would then be able to use all the output formats (latex, mathml, pretty, ...).

mmatera commented:
Well, I have spend some time studying how this is implemented in Wolfram Language, and I would say that something like `stringPict` and `prettyForm` is needed, but at a different level of the output generation. 

Sympy works a little bit differently from WL in the sense than here you have many different classes of expressions, while in WL, expressions are mostly all of the same type: a tree-like structure, with `Atom`-like terminal elements.  But at the level of formatting and printing, I guess the steps would be similar.

In Wolfram Mathematica (WMA), once an expression reach a "final" form, the interpreter applies some "format" rules. 
Format rules transforms the expression in a more "standard" formatting expression. For example,
`a+b[x]` in WL is represented as
 `Plus[a,b[x]]`.

 When Formatting is applied, it is converted into

`Infix[{a,b[x]},'+',310,None]` 

where the first argument is a list of terms, the second parameter is the operator that represents the operation, the forth is the "precedence" value, and the last one is the "associativity" (None==="both side associative"). At the end of the process, parenthesis are added acording to the precendece and associativity of the expression.

Then, the expression is converted to a "Box" form, that defines a subset of the language defining how an expression must be printed. For the example, the rules that apply to the expression in the "StandardForm" leads to

```
RowBox[{"a"," + ",  RowBox[{"b", "[" , "x" , "]" }]}]
```
Which can be easily converted to LaTeX, MathML or any other structured language, or reinterpreted (`MakeExpression`) to the original expression, and eventually "rendered"  into a pdf/web page/image file  or a ASCII terminal.

There is also an `OutputForm` which looks similar to what you are talking about, which produces a Boxed expression of the form
```
InterpretationBox[PaneBox["a + b[x]"], OutputForm[a+b[x]]]]
```
A `PaneBox` can contain multiline expressions, similar to the one produced by `pretty(...)`, and the `InterpretationBox` holds both the printable text and the expression that originates it. I need stringPict for this task.

This thing that you are proposing that combines a "PaneBox" (containing a stringPict) and a symbolic expression would be equivalent to the `InterpretationBox`, and I think it would be at a different level.

mmatera commented:
Regarding the methods included in stringPict / prettyPrint, I would expect that provide an interface to draw/render over,  and combine "text-mode window". 

Then, IMHO,  the logic regarding when an expression should be wrapped in parenthesis, and the connection between a part of the stringPict and the subexpression that produced it should be in a different level of abstraction. 

I found it a little bit strange that the `__add__` operator between two stringPicts produces a new stringPict with a "+" in the middle, and handles the logic of parenthesis. I would expect that it means block-wise concatenation (like `next` does, or `+` does for `str` objects), but OK, it does not bother too much.

mmatera commented:
> printed expressions. This would make reasoning about certain more complex printing situations easier because you could look inside what has already been printed. 

Regarding this, the way to handle it is to build expressions from outside to inside, so the drawer/printer knows the context of the object that is being rendered. At some points, `PrettyPrinter` tries to build the inner components and assemble them later, which is what produces this funny behavior. I think that improving the interface should make it clear how to handle the special cases.

asmeurer commented:
We should definitely consider whether it makes more sense for stringPict to become the representation of boxes (e.g., by gaining the ability to remember how it was built), or whether that should be a separate and new abstraction that is built on top of it (or separately from it). If the conclusion is the latter, then the question is how much we need to be refactoring stringPict itself instead. 

> I found it a little bit strange that the __add__ operator between two stringPicts produces a new stringPict with a "+" in the middle, and handles the logic of parenthesis. I would expect that it means block-wise concatenation (like next does, or + does for str objects), but OK, it does not bother too much.

I did not even realize this was the case. It's definitely trying to do too many things at once, or more specifically, it is trying to cover too many different levels of abstraction at the same time. It should either be low level, and not know anything about mathematics, just combining boxes together, or higher level and know something about how mathematical functions are printed. 

It seems like there are a lot of issues with stringPict. Even the name isn't really up to PEP 8 standards. It's a really old piece of code; one of the oldest in SymPy. So the question is whether we should just abandon it and rewrite it to something better, or, if we decide we don't consider it to be a public API, just completely refactor it and fix these issue. 

I like Oscar's way of framing it as it should be a sequence of conversions, like a compiler.

oscarbenjamin commented:
It sounds like the box form is what I'm thinking of. Basically an abstract representation of how the expression should be displayed but it should not have anything like stringPict because no strings should have been generated yet.

mmatera commented:
A way around would be to put this part of the code in a small library, and then hook both Mathics and Sympy from it. I can offer some help in the refactor, which goes parallel with the work I am doing in Mathics.

oscarbenjamin commented:
> A way around would be to put this part of the code in a small library, and then hook both Mathics and Sympy from it.

I don't think it is worth having a separate library for this. This PR can be merged but ultimately this part of SymPy should be reimplemented in a significantly different way.

I imagine that a refactor here would make something much more suitable for what Mathics would want. However I am not proposing that I will do that rewrite right now and I don't think Aaron is either so unless you want to do it then we can make these changes here now and Mathics can use it.

asmeurer commented:
Yes, I'm not trying to block the work on this PR, just trying to have a conversation about more significant work that could be done. 

I think the most important thing we need to decide here is just how public the stringPict API is. If we consider it to be nonpublic, then we should feel free to heavily refactor or even delete it. As long as the existing pretty printing tests continue to pass, or if the results do change they aren't an improvement. 

If we decide it is public API, then we should just leave it be and start a separate submodule that's independent of it. 

My understanding was that this PR does have some API breaks to stringPict (correct?), so it would be good to figure this out. 

If we feel that whatever we want will be completely different from stringPict anyways, then maybe the best thing would be to do the latter, since maintains it for people using it. We can decide if we want to [deprecate it](https://docs.sympy.org/latest/contributing/deprecations.html) once we have a usable replacement.

asmeurer commented:
https://github.com/search?type=code&auto_enroll=true&q=stringPict

It looks like galgebra uses stringPict https://github.com/pygae/galgebra/blob/327df0a1f9a2e338efa5ce373ce2abc1af0fc874/galgebra/atoms.py#L9. I didn't spot anything else major using it, just a few minor repos. But I think it's fair to say that *someone* out there depends on it. 

It's also mostly independent from the rest of SymPy so in principle someone could just copy the code if they still want to use it.

oscarbenjamin commented:
Would the changes here break galgebra and if so is it difficult to change them so that they don't break galgebra?

mmatera commented:
Looking at that code, it just need to change a line to work this this branch

oscarbenjamin commented:
It would be best if they don't need to change anything and their code still works.

rocky commented:
> Yes, exactly. Ideally the way that display printing should work is that an expression is first converted to a new type say DisplayExpression which represents how the expression should appear i.e. it symbolically represents things like brackets, superscripts etc. In the conversion you would have `Add -> DisplayAdd`, `Mul -> DisplayFraction/DisplayMul`, `Pow -> DisplaySurd/DisplayPow` etc. A function like `sin(x)` would be transformed to `DisplayCall('sin', x)` and so on. Then the various display printers like mathml, latex, pretty etc should operate on DisplayExpression and render the actual text. Then all Mathics would need to do is implement the transformation from a Mathics expression to DisplayExpression but would then be able to use all the output formats (latex, mathml, pretty, ...).

Actually, my understanding is that DisplayExpression in Wolfram Language lingo is called a BoxedExpression. IMHO all of this stuff is derivative from TeX's horizontal and vertical boxes and I think there are aspects in this in the CSS/HTML boxing model as well. But I am not a novice at this kind of stuff.

mmatera committed:
restoring parens as it was before. Adding parenthesis with the new behavior

mmatera commented:
15b7835 restores  `parens()` to is behavior in master, and rename `parenthesis()` the method with the new behavior. 
This seems to avoid breaking any of the projects listed in https://github.com/search?type=code&auto_enroll=true&q=stringPict

smichr commented:
> rename parenthesis() the method with the new behavior.

What about the more generic `bracket` which can refer to curly, square, angle or round shapes.

rocky commented:
> > rename parenthesis() the method with the new behavior.
> 
> What about the more generic `bracket` which can refer to curly, square, angle or round shapes.

I'd call this a "grouping symbol"

smichr commented:
```
pform.parenthesis(left="(", right=")")
pform.bracket(left="(", right=")")
pform.group(left="(", right=")")
pform.grouping(left="(", right=")")
```

If you are adding a new method, it would be nice to be able to just give both in one string, e.g. `pform.group('[]')`.

mmatera commented:
> ```
> pform.parenthesis(left="(", right=")")
> pform.bracket(left="(", right=")")
> pform.group(left="(", right=")")
> pform.grouping(left="(", right=")")
> ```
> 
> If you are adding a new method, it would be nice to be able to just give both in one string, e.g. `pform.group('[]')`.

Why? I think it makes the method less flexible, because you could want to put a bracket on one side and not on the other:
```
values = prettyForm("").above("1").below("0")
conditions = prettyForm("").above("x>1").below("x<1")
print(prettyForm("f(x)=").right(    values.right(" ").right(conditions).parenthesis(left="{",right="")))
```

On the other hand, there are symbols whose names are not characters. Suppose you want to draw the integer part of a fraction:
```
print(prettyForm(sympy.pretty(x/y)).parenthesis(left="lfloor",right="rfloor"))
```
to get
```
⎢x⎥
⎢─⎥
⎣y⎦
```

What would be possible is to add a third parameter that allows a shorthand notation, like `characters="lr"`,
in a way that if it is provided, use the first character for `left` and the second for `right`, but still keeping `left` and `right` as parameters.

rocky commented:
> ```
> pform.parenthesis(left="(", right=")")
> pform.bracket(left="(", right=")")
> pform.group(left="(", right=")")
> pform.grouping(left="(", right=")")
> ```
> 
> If you are adding a new method, it would be nice to be able to just give both in one string, e.g. `pform.group('[]')`.


> > ```
> > pform.parenthesis(left="(", right=")")
> > pform.bracket(left="(", right=")")
> > pform.group(left="(", right=")")
> > pform.grouping(left="(", right=")")
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > If you are adding a new method, it would be nice to be able to just give both in one string, e.g. `pform.group('[]')`.
> 
> Why? I think it makes the method less flexible, because you could want to put a bracket on one side and not on the other:

This now gets into the discussion of what "grouping", as contrasted to various operator forms impacts how one might want to think of boxing. 

For example, one might have a left prefix operator, a right prefix operator, or some unusual operator form as you have for integration where there are places for limits and a part before a partial symbol and what the derivative is with respect to.

Robert Jacobson completed a study of this for Mathematica. See https://wltools.github.io/LanguageSpec/Specification/Syntax/OperatorTableHTML/  .  His table I find a little bit overly complex but it contains  "Arity",  "Affix", and the "L", "O", and "N' columns indicate what appears on the left and right. For example, you might want a  Prefix or a Postfix kind of box. 

I have been slowly going over this to try to simplify and make it accessible programmatically from YAML in https://github.com/Mathics3/mathics-scanner/blob/0fd4fb69db50ea500348d408116ff1f46b6442f8/mathics_scanner/data/operators.yml



 Robert Jacobson has done an extensive study of this.

smichr commented:
> allows a shorthand notation, like `characters="lr"`,
> in a way that if it is provided, use the first character for `left` and the second for `right`, but still keeping `left` and `right` as parameters.

or 

Since this is not usually going to be something that is typed often from an interactive prompt, I guess `left` and `right` is fine. And the `floor` case is something that I hadn't considered.

oscarbenjamin commented:
> restores `parens()` to is behavior in master, and rename `parenthesis()` the method with the new behavior.
> This seems to avoid breaking any of the projects

To be clear would the galgebra code still work:
https://github.com/pygae/galgebra/blob/327df0a1f9a2e338efa5ce373ce2abc1af0fc874/galgebra/atoms.py#L142

mmatera commented:
> > restores `parens()` to is behavior in master, and rename `parenthesis()` the method with the new behavior.
> > This seems to avoid breaking any of the projects
> 
> To be clear would the galgebra code still work: https://github.com/pygae/galgebra/blob/327df0a1f9a2e338efa5ce373ce2abc1af0fc874/galgebra/atoms.py#L142

Sure, because `parens`, `next` and `stack` still do what they used to do.

Notice that if we define a new class from scratch (let's say, StrPicture) and use it in `PrettyPrinter`, then Galgebra would get broken, because it uses functions that return whatever `PrettyPrinter` generates.

oscarbenjamin commented:
> Notice that if we define a new class from scratch (let's say, StrPicture) and use it in `PrettyPrinter`, then Galgebra would get broken, because it uses functions that return whatever `PrettyPrinter` generates.

That is true but if we are breaking downstream code for a good reason, like because there is a major refactor to provide a much better interface then I consider that different. Also there could be ways to make it so that galgebra still works e.g. by not using `printer._print` to return that new type.

mmatera commented:
Looking again at the whole discussion, the comments related directly with this PR would be that

1.  [ ] It must avoid breaking downstream code (Galbebra).
2.  [ ] It must be picked good names for the new methods.
3.  [ ]  All the `sympy.printing.pretty` package should go under a full refactor, to make it more PEP8 compliant, and to improve its design. 

Regarding 1., by keeping `parens` as it was (returning a tuple), and adding a new method `parenthesis` that behaves as I had proposed `parens` should work (returning a stringPict object), the compatibility not broken.

Regarding 2. the only method that was in the discussion was `parenthesis`. (Notice that I also added `subindex`, `superindex`  and `subsuperindex`). The options for the identifier for `parenthesis` that were mentioned are :  "group", ,"grouping", "parenthesis", "wrap", "bracket", "embrace", "enclose"

I think that the criteria to choose one of them would be:
* The identifier is not used in a different context in the existing code.
* The identifier should suggest an action over a block of text/stringPict, not over an expression.
* The identified should not be too specific, since the method could be used to put different kind of parenthesis and brackets, ("()","[]", "{}","<>", "| |", "|| ||")  on both sides or just one side of the expression.
* The identified should be as short as possible.

With these criteria, IMHO:

* group / grouping: not sure if it suggests an operation over expressions instead of an operation over a block of text. 
* parenthesis: too long? too specific?
* wrap: already used in `sympy.polys` and `sympy.integrals.laplace`.
* bracket: maybe too specific? 
* enbrace, enclose:  ?

What do you guys think about?

mmatera commented:
> That is true but if we are breaking downstream code for a good reason, like because there is a major refactor to provide a much better interface then I consider that different. Also there could be ways to make it so that galgebra still works e.g. by not using `printer._print` to return that new type.

To make the change smooth, we could add a deprecation warning inside `parens`, to say that such a refactor is in progress.
Once the new methods are used in the downstream code, then we can just rename the class.

oscarbenjamin commented:
> Regarding 1., by keeping `parens` as it was (returning a tuple), and adding a new method `parenthesis` that behaves as I had proposed `parens` should work (returning a stringPict object), the compatibility not broken.

Geat, thanks.

> Regarding 2., I think it should go after the changes proposed here.

Agreed.

> Regarding 3. the only method that was in the discussion was `parenthesis`.

I think that `parenthesis` is fine if the existing method is called `parens`.

Ultimately I would expect that a rewritten version of how the display printers work would end up needing something different from stringPict anyway so it is probably not worth getting hung up on exactly the names of the methods. In the newer version you couldn't have methods like the current `.parens` or `.parenthesis` anyway because you wouldn't want eager bottom up rendering.

oscarbenjamin commented:
> To make the change smooth, we could add a deprecation warning inside `parens`, to say that such a refactor is in progress.
> Once the new methods are used in the downstream code, then we can just rename the class.

We should not add a deprecation warning before the non-deprecated replacement code is added. I would expect a compatibility interface to work differently anyway though: we can make it so that old code works after any refactor.

mmatera commented:
> > To make the change smooth, we could add a deprecation warning inside `parens`, to say that such a refactor is in progress.
> > Once the new methods are used in the downstream code, then we can just rename the class.
> 
> We should not add a deprecation warning before the non-deprecated replacement code is added. I would expect a compatibility interface to work differently anyway though: we can make it so that old code works after any refactor.

OK, maybe we should open a Discussion to define the steps of such a refactor, but what I have in mind is that something that converts generic *algebraic* expressions into a human-readable object  would be

1. Conversion from a Sympy (expression) object to a `Format` expression object (in WL, Infix/Prefix/Postfix/PrecedenceForm) which contains an abstract representation of precedence, and how to show a large class of operations).
2. "translate" these `Format` expressions into a typesetting language like MathML or LaTeX, or "render" to a binary or text picture. 

In WL, the second step is split in two: first translate it to `Box` expressions, which represent a sub-language for describing the typesetting, and then from that representation, produce MathML, LaTeX, or binary pictures. On the other hand, to produce textual output (InputForm and OutputForm) the `Format` expression is used to render an output text directly.

Here and in Mathics,  `stringPict` should accomplish the task of rendering the expression in an *ASCII-art*-like form. So, we still need it, holding certain functionality to position different text blocks. What we should change later is the interface between `PrettyPrinter` and `stringForm`. Galgebra uses Sympy objects to represent expressions, so it should interact with `PrettyPrinter`. Mathics does not use  Sympy objects, so we need to use `stringForm` directly.

mmatera committed:
trailing spaces

oscarbenjamin commented:
> maybe we should open a Discussion to define the steps of such a refactor,

Yes, we should do that in a separate discussion.

Does anyone have any views on the changes in this PR? To me they look good.

sylee957 commented:
Is the release notes updated? I notice that a lot of things had changed during the discussion and 

> the methods subscript and superscript were added to the stringPict interface

I can't see `subscript` and `superscript` method, but I can rather see `subindex`, `superindex`, `subsuperindices` added.

> the methods above, below, left, right and parens were modified to return prettyForm or stringPict
objects depending on the class of the object over which are called.

I think that `parens` did not modify the result type though. (you have decided to add `parenthesis`) 
But anyway, I notice that above/below/left/right changed to backward-incompatible types, so I'm still confused about whether you should change such things back or not.

I also think that the PR contains a bunch of formatting changes, but the things I find messy are
the changes from `# Add spacers between sub-pictures` to `#Add spacers between sub-pictures`.
If they are accidental, can you clean up such changes?

I also think that there are some commented-out codes and added TODO comments. So I'm unclear about whether the work is complete for merge.

mmatera commented:
> Is the release notes updated? I notice that a lot of things had changed during the discussion and I notice that some things are outdated
> 

No, they aren't. I was waiting to converge in the name of the methods and possible other changes, observations.

> > the methods subscript and superscript were added to the stringPict interface
> 
> I can't see `subscript` and `superscript` method, but I can rather see `subindex`, `superindex`, `subsuperindices` added.
> 
> > the methods above, below, left, right and parens were modified to return prettyForm or stringPict
> > objects depending on the class of the object over which are called.
> 
> I think that `parens` did not modify the result type though. (you have decided to add `parenthesis`) But anyway, I notice that above/below/left/right changed to backward-incompatible types, so I'm still confused about whether you should change such things back or not.
> 

Yep, I reverted it to ensure backward compatibility with the packages that use this module. Methods like `above`/`below`/`left`/`right` where not used outside, essentially because does not provide any extra functionality/advantage over the lower level methods `stack`/`next`. `parens` is used outside, so I keep it as it was before, and added the new method `parenthesis` that returns `stringPict` objects instead of tuples.

> I also think that the PR contains a bunch of formatting changes, but the things I find messy are the changes from `# Add spacers between sub-pictures` to `#Add spacers between sub-pictures`. If they are accidental, can you clean up such changes?
> 

I used a linter (black) to look for flackiness in the code, and it made these changes. I reverted some of them, but it seems some others prevailed. 

> I also think that there are some commented-out codes and added TODO comments. So I'm unclear about whether the work is complete for merge.

Once I get some degree of consensus on how the changes should go/corrected, I will make the changes and update the release notes.

mmatera committed:
removing trivial changes and commented out code

sylee957 commented:
I think that some backward incompatible changes to method like `above, below, left, right` can be unnecessary.
The only necessary change you may want is that your custom subclass of `stringPict` want to override the calls to `stringPict` or `prettyForm`.
However, you can achieve similar thing like below:

```
class stringPict:
    def string_pict(self, s, baseline=0):
        return stringPict(s, baseline)
    ...

class prettyForm(stringPict):
    def string_pict(self, s, baseline=0):        
        return prettyForm(s, baseline)
    ...

class CustomStringPict(stringPict):
    def string_pict(self, s, baseline=0):        
        return CustomStringPict(s, baseline)
```

you can use to replace things like `arg.parenthesis()` to `arg.string_pict(*arg.parens())`.

I think that you have added a lot of construction like `self.__class__(...)`, but my general understanding is that it is always possible to decompose that into functions that returns constructor arguments and the function that constructs the object.
And if that makes the things more backward compatible, it can be favorable.

Maybe [self type](https://peps.python.org/pep-0673/) outputs like `self.__class__(...)` can be more useful for the design eventually, however, although [self type](https://peps.python.org/pep-0673/) is used quite commonly in Python community, however, my study is not very deep about that to give opinion about the design.
I note that self type is studied as second order logic for object oriented programming (such as https://link.springer.com/chapter/10.1007/978-1-4419-8598-9_16) which is already very involved concept in deep.

mmatera commented:
> I think that some backward incompatible changes to method like `above, below, left, right` can be unnecessary. The only necessary change you may want is that your custom subclass of `stringPict` want to override the calls to `stringPict` or `prettyForm`. However, you can achieve similar thing like below:
> 
@sylee957, Indeed, I can also avoid completely using stringPict by defining my own class. What I am proposing here are some changes that allow to producing an output as close as possible to the `pretty()` from Sympy, using another class hierarchy (Mathics expressions). As it is written now, despite there being a kind of abstraction hierarchy between stringPict and PrettyPrinter, the code is still quite entangled. So, using it from outside is almost as hard as doing it from scratch. So, in practice, PrettyPrinter only works with Sympy objects. But worst, the implementation of the PrettyPrinter also results unnecessarily complicated, and full of code duplication. 

> ```
> class stringPict:
>     def string_pict(self, s, baseline=0):
>         return stringPict(s, baseline)
>     ...
> 
> class prettyForm(stringPict):
>     def string_pict(self, s, baseline=0):        
>         return prettyForm(s, baseline)
>     ...
> 
> class CustomStringPict(stringPict):
>     def string_pict(self, s, baseline=0):        
>         return CustomStringPict(s, baseline)
> ```
> 
> you can use to replace things like `arg.parenthesis()` to `arg.string_pict(*arg.parens())`.

Yep, I am well aware of what Inheritance and Polymorphism mean. My goal was just to make that my subclasses play well together with the rest of the code.

> 
> I think that you have added a lot of construction like `self.__class__(...)`, but my general understanding is that it is always possible to decompose that into functions that returns constructor arguments and the function that constructs the object. 
And if that makes the things more backward compatible, it can be favorable.
> 

The point is that nobody used these methods outside of this project, because does not provide an improvement over the low-level methods.

> Maybe [self type](https://peps.python.org/pep-0673/) outputs like `self.__class__(...)` can be more useful for the design eventually, however, although [self type](https://peps.python.org/pep-0673/) is used quite commonly in Python community, however, my study is not very deep about that to give opinion about the design. I note that self type is studied as second order logic for object oriented programming (such as https://link.springer.com/chapter/10.1007/978-1-4419-8598-9_16) which is already very involved concept in deep.

OK, maybe @rocky have something to say about this. I just find it a compact and simple solution to keep compatibility and avoid some complicated idioms like  [target_type](*object.method_producing_arguments()) .

oscarbenjamin commented:
> As it is written now, despite there being a kind of abstraction hierarchy between stringPict and PrettyPrinter, the code is still quite entangled. So, using it from outside is almost as hard as doing it from scratch. So, in practice, PrettyPrinter only works with Sympy objects. But worst, the implementation of the PrettyPrinter also results unnecessarily complicated, and full of code duplication.

This is true although the best way to fix it would be by having a separate display expression type. That eliminates the duplication and makes it better for use from Mathics. Then it is possible to share the use of more complicated code than just the stringPict methods. If we are making bigger changes here then I would suggest to do it by not using stringPict and prettyForm at all. They can be left as is for backwards compatibility if downstream code is using them.

mmatera commented:
> This is true although the best way to fix it would be by having a separate display expression type. That eliminates the duplication and makes it better for use from Mathics. Then it is possible to share the use of more complicated code than just the stringPict methods. If we are making bigger changes here then I would suggest to do it by not using stringPict and prettyForm at all. They can be left as is for backwards compatibility if downstream code is using them.

OK, suppose that we add a brand new class StringBlock, that does what I put here in stringPict. Then, as soon as I use it in PrettyPrint, backward compatibility is fully broken. 

The other possibility is to start a separate `sympy.printing.pretty` module (let's say, `sympy.printing.pretty2`). But such a large project goes beyond what I would do alone. Again, my goal with this was just to provide a solution useful for both projects and in a way that makes it easier to produce similar outputs.

oscarbenjamin commented:
I don't understand why there is any need to break any compatibility for anything here.

Basically you don't like the exact behaviour of some of the stringPict methods and would prefer something different: just add some new methods to stringPict.

Am I missing something about why that doesn't work?

The diff here contains lots of changes that I presume are not really necessary which makes it harder to see what are the things that actually do need to change. I don't object to improving the code if there is a new method that makes the code simpler or better or whatever. I do prefer though if changing something in an incompatible way is done separately from other changes (i.e. not mixed into a larger than necessary diff).

mmatera commented:
@oscarbenjamin, my proposal here was to slightly modify the internal interface of stringPict in a way that PrettyPrint and other projects can use a higher-level interface to draw equations. Then, if PrettyPrint (which only works with the Sympy hierarchy) does not have to deal with finding the baseline for drawing a superindex or the size and position of a radical or an integration symbol, but uses that high level interface, then any other application that uses such an interface would get results which looks similar. This is something that I cannot achieve by using a subclass. 

Actually, I have already written a class that does more or less the same. However,  what we have discussed with @rocky is that building another library that is essentially the same would be a waste of resources. So I was looking at what would be necessary to hook the existing code in `sympy.printing.pretty` in a way that both projects could play and be maintained together. 

Of course, it just makes sense if we agree on what that interface should be.

oscarbenjamin commented:
> my proposal here was to slightly modify the internal interface of stringPict in a way that PrettyPrint and other projects can use a higher-level interface to draw equations.

Okay so why does a "slight change" come with a 1000 line diff?

> This is something that I cannot achieve by using a subclass.

We can just add some new methods to stringPict though without needing to change hundreds of lines of code and without breaking compatibility for any downstream code.

> However, what we have discussed with @rocky is that building another library that is essentially the same would be a waste of resources.

I agree but I think that there can be a significantly better level of abstraction for sharing code here than the stringPict and prettyForm classes. If we add a few new methods to stringPict and then build everything in mathics from there then there will still be a lot of duplicated code, especially if you consider the other printers like mathml, LaTeX etc. 

The proper solution here is to have something like the display form expressions as an intermediate representation and then something that renders those display forms to different kinds of string representations. We would want that in SymPy anyway and it would mean using all new code for all of these things. In that situation I think that stringPict and prettyForm would not be used any more and would exist only for some backwards compatibility interface for downstream code. There would naturally be some code that still does the sort of things that stringPict does but it would be a small part of the whole and would not have a similar interface.

mmatera commented:
> 
> Okay so why does a "slight change" come with a 1000 line diff?
>
Because the "slight change" affects many lines, but it does always in the same way.
 
> We can just add some new methods to stringPict though without needing to change hundreds of lines of code and without breaking compatibility for any downstream code.
> 

OK, when I have time I can rework this defining new methods. In any case, the "hundres of lines" change are internal changes: the only thing that "breaks" the interface would be the redefinition of `left`, `right`,  `top` and `bottom` that seem not to be used outside Sympy. In any case, I can undo the changes, and rewrite these methods with new names, to avoid this "massive" change.

> > However, what we have discussed with @rocky is that building another library that is essentially the same would be a waste of resources.
> 
> I agree but I think that there can be a significantly better level of abstraction for sharing code here than the stringPict and prettyForm classes. If we add a few new methods to stringPict and then build everything in mathics from there then there will still be a lot of duplicated code, especially if you consider the other printers like mathml, LaTeX etc.
> 

I didn't look in detail at the implementation of the other printers. And probably different printers need different intermediate format rules. But yes, sure.

> The proper solution here is to have something like the display form expressions as an intermediate representation and then something that renders those display forms to different kinds of string representations. We would want that in SymPy anyway and it would mean using all new code for all of these things. In that situation I think that stringPict and prettyForm would not be used any more and would exist only for some backwards compatibility interface for downstream code. There would naturally be some code that still does the sort of things that stringPict does but it would be a small part of the whole and would not have a similar interface.

OK, but then probably the better is to close this PR, build it from outside, and finally try to integrate it into the existing code. In any case, all this discussion round helps to understand better what are the design patterns that you want to keep.

rocky commented:
> The proper solution here is to have something like the display form expressions as an intermediate representation and then something that renders those display forms to different kinds of string representations. We would want that in SymPy anyway and it would mean using all new code for all of these things. In that situation I think that stringPict and prettyForm would not be used any more and would exist only for some backwards compatibility interface for downstream code. There would naturally be some code that still does the sort of things that stringPict does but it would be a small part of the whole and would not have a similar interface.

I agree. Again, in WMA this is called Boxed Expressions. However the idea of Boxing was first developed in TeX with its horizontal and vertical boxes and glue, and boxing concepts exist in the CSS Box Model. TeX and CSS are examples of low-level boxing.  But on top of that,  one might have high-levels of boxing that match the generic forms found in continued fractions, square roots, integrals, etc. I think that's what WMA does.

Personally, I am happier to wait for a better more general solution that works across SymPy, Mathics3 and other things that might use SymPy than a temporary quick fix. On the Mathics side, [our roadmap was to work on the Boxing aspect](https://github.com/Mathics3/mathics-core/blob/master/FUTURE.rst#2023-roadmap) of this first to get that more solid, and then worry about the Boxing to formatting aspect.

sylee957 commented:
I'm okay with the changes. However, the only thing that we have hard time following the PR is that your statements have [temporality](https://en.wikipedia.org/wiki/Temporal_logic).

The obvious thing is that once we merge the PR, `mathics` starts depending heavily on `stringPict`, and we have much stronger constraint for making backward incompatible changes further. So the statements like:

> The point is that nobody used these methods outside of this project, because does not provide an improvement over the low-level methods.
> So I was looking at what would be necessary to hook the existing code in sympy.printing.pretty in a way that both projects could play and be maintained together.

is not invariant for the time or consequences of merging this PR. And we may have risk factors, for example, committing some premature ideas, so that was one of the reason why I had taken some look at the PR. But I'm not entirely against the idea of this PR if I had questions answered at least.

There are also other references that we can use to design presentational form of equations.

For example, the MathML [documentation](https://developer.mozilla.org/en-US/docs/Web/MathML) can be useful typesetting of mathematics, in reasonable level of abstraction. And I think that MathML is easier to study than TeX for example, because there are quite a few special presentation elements (such as `<mfrac>, <msqrt>, <mtr>, ...`) and we can easily figure out that we may only need a few elements like that for printing general mathematical expressions in most useful way.

There are also other relevant topics like [multidimensional stringology](https://www.stringology.org/papers/Zdarek-PhD_thesis-2010.pdf), array grammar, [picture languages](https://doi.org/10.1016/C2013-0-11407-9) that can be useful for implementing ASCII-art, but I am not expert about that, so it's up for you to investigate.
